File: week10/burning-airlines-immersive/ba-rails-backend/db/seeds.rb

User.destroy_all
print "Creating users... "
u1 = User.create! name: 'Test User 1', email: 'one@one.com', password: 'chicken'
u2 = User.create! name: 'Test User 2', email: 'two@two.com', password: 'chicken'
u3 = User.create! name: 'Test User 3', email: 'three@three.com', password: 'chicken'
puts "created #{User.count} users."

Airplane.destroy_all
print "Creating airplanes... "
a1 = Airplane.create! name: '737', rows: 40, cols: 6
a2 = Airplane.create! name: '737 MAX', rows: 80, cols: 8
puts "created #{Airplane.count} airplanes."

Flight.destroy_all
print "Creating flights... "
f1 = Flight.create! flight_number: 'BA256', departure_date: '2021-10-01 04:20', origin: 'SYD', destination: 'MEL', airplane_id: a1.id
f2 = Flight.create! flight_number: 'BA512', departure_date: '2021-10-02 11:20', origin: 'SYD', destination: 'MEL', airplane_id: a1.id
f3 = Flight.create! flight_number: 'BA1024', departure_date: '2021-10-03 16:20', origin: 'SYD', destination: 'SIN', airplane: a2
puts "created #{Flight.count} flights"

Reservation.destroy_all
print "Creating reservations... "
r1 = Reservation.create! flight_id: f1.id, user_id: u1.id, row: 10, col: 2
r2 = Reservation.create! flight: f1, user: u2, row: 10, col: 3
r3 = Reservation.create! flight: f1, user: u3, row: 10, col: 4
r4 = Reservation.create! flight: f2, user: u3, row: 15, col: 1
puts "created #{Reservation.count} reservations."

puts
puts "Association tests:"
puts "Airplane #1 has #{Airplane.first.flights.length} flights (should be 2)"
# You can automatically create an error if you don't get the right answer here!
# (Like JS "throw new Error('abc')")
raise Exception.new('Airplane flights association broken') if Airplane.first.flights.length != 2
# expect( Airplane.first.flights.length ).to eq 2


puts "Flight #1 has #{Flight.first.reservations.length} reservations (should be 3)"
puts "User #3 has #{User.third.reservations.length} reservations (should be 2)"

puts "Flight #1 has airplane name #{Flight.first.airplane.name} (should be '737')"
puts "Reservation #1 has user name #{Reservation.first.user.name} (should be 'Test User 1')"
puts "Reservation #1 has flight number #{Reservation.first.flight.flight_number} (should be 'BA256')"

puts
puts "Done."

File: week10/node-servers/ba-express-mongodb-backend/models/seeds.js

const bcrypt = require('bcrypt');


// flight seed data
const flights =  [
  {
    flight_number: 'BA123',
    origin: 'SYD',
    destination: 'MEL',
    departure_date: new Date('2021-10-01T04:20:00Z'),
    airplane: { name: '737', rows: 10, cols: 4  },
    reservations: [
      { row: 1, col: 1, user_id: 10 },
      { row: 2, col: 2, user_id: 10 },
      { row: 3, col: 3, user_id: 11 },
    ] // reservations
  }, // first flight
  {
    flight_number: 'BA456',
    origin: 'SYD',
    destination: 'MEL',
    departure_date: new Date('2021-11-01T04:20:00Z'),
    airplane: { name: '767', rows: 12, cols: 6  },
    reservations: [
      { row: 1, col: 1, user_id: 10 },
      { row: 2, col: 2, user_id: 10 },
      { row: 3, col: 3, user_id: 11 },
    ] // reservations
  }, // second flight
  {
    flight_number: 'BA789',
    origin: 'SYD',
    destination: 'SIN',
    departure_date: new Date('2021-11-01T04:20:00Z'),
    airplane: { name: '767', rows: 12, cols: 6  },
    reservations: [
      { row: 1, col: 1, user_id: 10 },
      { row: 2, col: 2, user_id: 10 },
      { row: 3, col: 3, user_id: 11 },
    ] // reservations
  }, // third flight
];


const mongoose = require('mongoose');

// Load our model files
const Flight = require('./Flight');
const User = require('./User');

// Connect to the DB
mongoose.connect('mongodb://localhost/ba', { useNewUrlParser: true, useUnifiedTopology: true });

const db = mongoose.connection;

db.on('error', (err) => {
  console.log('Connection error', err);
});

// Once the connection is established, we can start querying/seeding
db.once('open', async () => {

  // Flight.find()
  // .then( data => {
  //   console.log('Flights:', data);
  // })
  // .catch( console.warn );
  //
  // Seed from here
  // 1. await deleting flights:
  //    await Flight.deleteMany( {} );      // destroy_all
  //    await Flight.create( [ {}, {} ] )
  //
  //
  // Create the Burning Airlines API routes in a server.js express file!
  // Use the same Mongoose connection lines as in this file!
  //
  // Remember to wrap your 'await' code in try-catch blocks to catch errors!

  const flightsCreated = await createFlights();
  console.log('Flights done, created ', flightsCreated.length );

  // Now move on to creating users
  const usersCreated = await createUsers( flightsCreated ); // pass the created flights in
  console.log('Users done, created ', usersCreated.length );
  console.log(usersCreated[0]);
  console.log(usersCreated[0].reservations);

  process.exit(0); // quit program

}); // once connection is established


const createFlights = async () => {

  try {

    // 1. Empty the flights collection (table), just like Rails Flight.destroy_all
    await Flight.deleteMany();

    // 2. Insert the new flights
    await Flight.create( flights ); // array of flights

    const flightResults = await Flight.find();
    // console.log('flights', flightResults);

    return flightResults; // this will be the resolved promise value of this async function

  } catch(err){
    console.log('Error creating flights:', err);
  }

}; // createFlights


const createUsers = async (flights) => {

  // user seed data


  const users = [
    {
      name: 'Test User 1',
      email: 'one@one.com',
      passwordDigest: bcrypt.hashSync('chicken', 10),
      // reservations?? need to know flight IDs first!
      reservations: [
        {
          row: 1, col: 1,
          flight: flights[0]
        },
        {
          row: 2, col: 2,
          flight: flights[1]
        },
      ], // reservations
    },
    {
      name: 'Test User 2',
      email: 'two@two.com',
      passwordDigest: bcrypt.hashSync('chicken', 10),
      // reservations?? need to know flight IDs first!
      reservations: [
        {
          row: 3, col: 3,
          flight: flights[0]._id
        },
        {
          row: 4, col: 4,
          flight: flights[1]._id
        },
      ], // reservations
    },
  ];


  try {

    await User.deleteMany(); // clear the collection! .destroy_all

    await User.create( users ); // TODO: no reservations!

    const createdUsers = await User.find().populate('reservations.flight');

    return createdUsers;

  } catch(err) {
    console.log('Error creating users', err);
  }

}; // createUsers
